exams/pearson/audit.py

```diff
@@ -0,0 +1,265 @@
+    """
+    Encrypted file auditor for exam requests/responses
+    """
+    REQUEST = 'request'
+    RESPONSE = 'response'
+
+    def __init__(self, gpg=None, store=None):
+        self.gpg = gpg or gnupg.GPG()
+        self.store = store or S3AuditStorage()
+
+    @property
+    def fingerprint(self):
+        """The fingerprint for the public GPG key"""
+        return settings.EXAMS_AUDIT_ENCRYPTION_FINGERPRINT
+
+    @property
+    def public_key(self):
+        """The public GPG key"""
+        # unescape newlines (from an env var)
+        return re.sub(r'\\\\n', '\n', settings.EXAMS_AUDIT_ENCRYPTION_PUBLIC_KEY)
+
+    def get_fingerprints(self):
+        """
+        Get the fingerprints of keys currently loaded on the system
+
+        Returns:
+            list:
+                list of keys
+        """
+        return self.gpg.list_keys().fingerprints
+
+    def configure(self):
+        """
+        Performs a one-time configuration of GPG keys
+
+        Raises:
+            ImproperlyConfigured:
+                settings are missing or invalid
+
+        Returns:
+            bool:
+                True if configuration successful
+        """
+        # if any of the settings are not set, we consider this feature not enabled
+        if any(getattr(settings, key) is None for key in REQUIRED_SETTINGS):
+            raise ImproperlyConfigured(
+                "One of EXAMS_AUDIT_ENCRYPTION_PUBLIC_KEY or "
+                "EXAMS_AUDIT_ENCRYPTION_FINGERPRINT is required but not set"
+            )
+
+        if self.fingerprint in self.get_fingerprints():
+            return
+
+        result = self.gpg.import_keys(self.public_key)
+
+        counts = dict(result.counts)
+
+        if counts['not_imported'] != 0:
+            raise ImproperlyConfigured('Error importing GPG keys for exam auditing: {}'.format(result.summary()))
+
+        if self.fingerprint not in result.fingerprints:
+            raise ImproperlyConfigured(
+                'Exam auditing GPG key successfully imported but does not match fingerprint: {}'.format(
+                    self.fingerprint
+                )
+            )
+
+    @contextmanager
+    def encrypted_file(self, filename):
+        """
+        Encrypt the file locally
+
+        Args:
+            filename:
+                absolute path to the local file
+
+        Returns:
+            str:
+                absolute path to the local encrypted file
+        """
+        encrypted_filename = '{}.gpg'.format(filename)
+
+        log.debug('Encrypting file %s to %s', filename, encrypted_filename)
+
+        with open(filename) as source_file:
+            self.gpg.encrypt(
+                source_file,
+                self.fingerprint,
+                armor=False,
+                symmetric=False,
+                output=encrypted_filename
+            )
+
+        try:
+            yield encrypted_filename
+        finally:
+            self._cleanup(encrypted_filename)
```

Comment:
If you want to make sure that `_cleanup()` is called to clean up any encrypted file that might have been generated, wouldn't it be better to put the `gpg.encrypt()` call inside the `try` block? The way this code is currently written, there could be an exception in that call after the file is already written, and `_cleanup()` wouldn't get called.